### DISCUSSION TITLE
Do moon repo offer concurrency?

### DISCUSSION THREAD

**Discussion by diraneyya** [moonrepo/moon#815] [Q&A] [answered]
Hey all 👋

As we continue to adopt moon as the official monorepo tool my in my full-stack development cohort, we ran into an issue using the `concurrently` package.

Not to use strong language, but  a package such as concurrently should not exist, rather, a decent task runner should instead offer this functionality (let me know what you think of this).

Motivated by an error using concurrently (in which the commands `tsc` and `nodemon` were not found) me and my students decided to aim for using moon itself for concurrent execution. As it turns out, this is very easily and beautifully done using:
```bash
moon run server:dev-tsc server:dev-nodemon
```

We also liked the consistency of the output formatting, compared to the `[0]` and `[1]` markers of concurrently, all achieved without any configuration.

However, we weren't so sure as we went from there to investigating the right way to enter the concurrent task into the tasks file in moon, allowing us to invoke the above using the following command:
```bash
moon run server:dev
```

Doing so we tried using a `noop` command with the `deps` property and using the task references of `~:dev-tsc` and `~:dev-nodemon`. It was a clean syntax and we hoped it would work, but it did not.

My understanding is that the `deps` property is intended for _build dependencies_ and not for concurrent execution, hence it is likely that one of the _watching_ command was blocking the rest of the pipeline.

As a result we ended up using the following:
```yaml
tasks:
  dev:
    command: "moon run server:dev-tsc server:dev-nodemon"
```

I do not like this for a few reasons. The one thing that stands out the most is the use of the redundant scope of `server` instead of self `~`. The other thing is mentioning the `moon` command itself, which also seems redundant.

I am posting to see what you think of this, if there is a better way to achieve the same, and hopefully have more fun and constructive discussions around this and related wishes of a modern task runner.

**Comment by milesj** [comment, marked as answer]
Hey @diraneyya,

Great question. A `noop` command with `deps` is our go to solution for this at the moment, here's an example: https://moonrepo.dev/docs/cheat-sheet#run-multiple-watchersservers-in-parallel

You mentioned you tried this. Can you talk about why it didn't work? It should theoretically be the same as `moon run server:dev-tsc server:dev-nodemon`.

**Reply by diraneyya** [reply]
Thank you for the reply!

I tested it again and it indeed works 🙈 I am not sure why my earlier test showed the opposite. Apologies!

I like this syntax much better, but the output of the dependencies seem to be suppressed.

In an attempt to reproduce the same output as with `moon run server:dev-tsc server:dev-nodemon` I used the following (which did not work):
```yaml
  dev:
    command: "noop"
    deps:
      - "~:dev-tsc"
      - "~:dev-nodemon"
    options:
      outputStyle: 'stream'
```

Is there a way to produce similar output from the dependencies as the case is when running multiple tasks?

**Reply by milesj** [reply]
@diraneyya Are you saying the output of `dev-tsc` and `dev-nodemon` are not showing? Or the deps of those deps?

Be sure you're `dev-*` tasks have `local: true` or the `outputStyle` option set. `outputStyle` only applies to the current task and not its deps.

**Reply by diraneyya** [reply]
Oh I see! I got to reading the following, which clarified things to me:
<img width="734" alt="image" src="https://user-images.githubusercontent.com/107997002/234938287-456b595e-ef43-40cc-a130-3b3dcfa2c402.png">

I did not realise that by switching the task's name from `dev` to `dev-tsc` or `dev-nodemon` that I was _implicitly_ affecting the `outputStyle` setting, implied by the `local` setting 🤯🙈🤓

I have explicitly specified the `outputStyle` of both constituent tasks to `'stream'` and now I can see the output from the dependencies. And yes it makes sense that `outputStyle` and other options apply to the task itself, rather than its dependencies.

Thank you!

**Reply by milesj** [reply]
Awesome, glad it works 👍
